contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,215 @@
+    pub pool_size: u32,
+    /// Any extra options that are included in the configuration.
+    pub extras: Option<&'a BTreeMap<String, Value>>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
+    let database_configs = match from.get_slice("databases") {
+        Ok(configs) => configs,
+        Err(_) => {
+            error_!("Rocket.toml does not contain any database configurations when attempting to fetch database: {}", name);
+            return None;
+        }
+    };
+
+    let connection_config = match database_configs
+        .iter()
+        .filter(|e| {
+            match e.as_table().map(|table| {
+                table.get("name").map(|v| v.as_str()) == Some(Some(name))
+            }) {
+                Some(true) => true,
+                _ => false,
+            }
+        })
+        .nth(0) {
+        Some(config) => config,
+        None => {
+            error_!("Could not find any database configurations named: {}", name);
+            warn_!("Available configurations: {}", available_database_configs(from).join(", "));
+            return None;
+        }
+    };
+
+    let url = match connection_config.get("url") {
+        Some(url) => {
+            match url.as_str() {
+                Some(url) => url,
+                None => {
+                    error_!("Could not convert the URL into a valid string");
+                    return None;
+                }
+            }
+        }
+        None => {
+            error_!("Could not fetch the URL as part of the connection config");
+            return None;
+        }
+    };
+
+    let pool_size = match connection_config.get("pool_size") {
+        Some(pool_size) => {
+            pool_size.as_integer().unwrap_or(from.workers as i64) as u32
+        }
+        None => from.workers as u32,
+    };
+
+    let extras = connection_config.as_table();
+
+    Some(DatabaseConfig {
+        url,
+        pool_size,
+        extras,
+    })
+}
+
+/// Returns an iterator containing the names of all the possible database connections.
+/// The primary use within `rocket_contrib` for this method is in relation to calling out
+/// valid database connections that are specified in the `Rocket.toml` file.
+pub fn available_database_configs<'a>(config: &'a config::Config) -> Vec<&'a str> {
+    let db_config = match config.get_slice("databases") {
```

Comment:
This is just: `config.get_slice("databases").unwrap_or(vec![]);`

Reply:
Yes, it is. Thanks for catching this.